File: video/api/build_project.py
"""Parameterized OpenCut project builder (importable).

A function-shaped version of build_opencut_project.py: takes audio + transcript +
images dir as arguments instead of module-level constants, so the API worker can
assemble any job's artifacts. The standalone script stays as-is for manual use.
"""
from __future__ import annotations

import json
import os
import re
import shutil
import struct
import uuid
import wave

TICKS_PER_SECOND = 120_000
CANVAS = {"width": 1920, "height": 1080}
NS = uuid.UUID("6f0c1e00-0000-4000-8000-000000000001")  # fixed namespace → stable ids

_IMG_RE = re.compile(r"^(\d+)_(\d\d)-(\d\d)\.png$")
_CAP_RE = re.compile(r"^\[(\d{1,2}):(\d{2})\]\s*(.+?)\s*$")


def _stable_id(name: str) -> str:
    return str(uuid.uuid5(NS, name))


def _ticks(seconds: float) -> int:
    return round(seconds * TICKS_PER_SECOND)


def _png_size(path: str):
    with open(path, "rb") as f:
        head = f.read(24)
    w, h = struct.unpack(">II", head[16:24])
    return w, h


def _wav_duration(path: str) -> float:
    with wave.open(path, "rb") as w:
        return w.getnframes() / float(w.getframerate())


def _srt_time(seconds: float) -> str:
    ms = int(round(seconds * 1000))
    h, ms = divmod(ms, 3_600_000)
    m, ms = divmod(ms, 60_000)
    s, ms = divmod(ms, 1000)
    return f"{h:02d}:{m:02d}:{s:02d},{ms:03d}"


def build_project(
    audio_file: str,
    images_dir: str,
    out_path: str,
    transcript_file: str | None = None,
    name: str = "Generated Video",
    include_captions: bool = False,
) -> dict:
    """Write `<out_path>` (.opencut.json) plus a sibling .srt. Returns a summary dict."""
    audio_dur_s = _wav_duration(audio_file)
    audio_dur_ticks = _ticks(audio_dur_s)
    audio_name = os.path.basename(audio_file)
    audio_media_id = _stable_id(audio_name)

    imgs = []
    for n in sorted(os.listdir(images_dir)):
        m = _IMG_RE.match(n)
        if not m:
            continue
        mm, ss = int(m.group(2)), int(m.group(3))
        imgs.append({"name": n, "start_s": mm * 60 + ss})
    if not imgs:
        raise ValueError(f"no timestamped PNGs (NNN_MM-SS.png) found in {images_dir}")

    image_elements = []
    manifest = [{
        "mediaId": audio_media_id,
        "filename": audio_name,
        "type": "audio",
        "size": os.path.getsize(audio_file),
        "duration": round(audio_dur_s, 3),
    }]
    for i, img in enumerate(imgs):
        start_s = img["start_s"]
        end_s = imgs[i + 1]["start_s"] if i + 1 < len(imgs) else audio_dur_s
        start_t = _ticks(start_s)
        dur_t = _ticks(end_s) - start_t
        mid = _stable_id(img["name"])
        w, h = _png_size(os.path.join(images_dir, img["name"]))
        image_elements.append({
            "id": _stable_id("el-" + img["name"]),
            "name": img["name"],
            "type": "image",
            "mediaId": mid,
            "startTime": start_t,
            "duration": dur_t,
            "trimStart": 0,
            "trimEnd": 0,
            "transform": {"scaleX": 1, "scaleY": 1, "position": {"x": 0, "y": 0}, "rotate": 0},
            "opacity": 1,
        })
        manifest.append({
            "mediaId": mid,
            "filename": img["name"],
            "type": "image",
            "size": os.path.getsize(os.path.join(images_dir, img["name"])),
            "width": w,
            "height": h,
        })

    caps = []
    if transcript_file and os.path.exists(transcript_file):
        with open(transcript_file) as f:
            for line in f:
                m = _CAP_RE.match(line)
                if m:
                    caps.append({"start_s": int(m.group(1)) * 60 + int(m.group(2)), "text": m.group(3)})

    caption_elements = []
    if include_captions:
        for i, cap in enumerate(caps):
            start_s = cap["start_s"]
            end_s = caps[i + 1]["start_s"] if i + 1 < len(caps) else audio_dur_s
            start_t = _ticks(start_s)
            dur_t = max(_ticks(end_s) - start_t, 1)
            caption_elements.append({
                "id": _stable_id(f"cap-{i}"),
                "name": cap["text"][:40],
                "type": "text",
                "content": cap["text"],
                "fontSize": 6,
                "fontFamily": "Arial",
                "color": "#ffffff",
                "background": {
                    "enabled": True, "color": "rgba(0,0,0,0.6)", "cornerRadius": 8,
                    "paddingX": 24, "paddingY": 12, "offsetX": 0, "offsetY": 0,
                },
                "textAlign": "center",
                "fontWeight": "bold",
                "fontStyle": "normal",
                "textDecoration": "none",
                "letterSpacing": 0,
                "lineHeight": 1.2,
                "startTime": start_t,
                "duration": dur_t,
                "trimStart": 0,
                "trimEnd": 0,
                "transform": {"scaleX": 1, "scaleY": 1, "position": {"x": 0, "y": 330}, "rotate": 0},
                "opacity": 1,
            })

    audio_element = {
        "id": _stable_id("el-audio"),
        "name": audio_name,
        "type": "audio",
        "sourceType": "upload",
        "mediaId": audio_media_id,
        "volume": 1,
        "startTime": 0,
        "duration": audio_dur_ticks,
        "trimStart": 0,
        "trimEnd": 0,
        "sourceDuration": audio_dur_ticks,
    }

    now = "2026-06-22T00:00:00.000Z"
    scene_id = _stable_id("scene-main")
    project_id = _stable_id("project-" + name)
    scene = {
        "id": scene_id,
        "name": "Main scene",
        "isMain": True,
        "tracks": {
            "overlay": ([{
                "id": _stable_id("track-captions"),
                "name": "Captions",
                "type": "text",
                "elements": caption_elements,
                "hidden": False,
            }] if include_captions and caption_elements else []),
            "main": {
                "id": _stable_id("track-main"),
                "name": "Main Track",
                "type": "video",
                "elements": image_elements,
                "muted": False,
                "hidden": False,
            },
            "audio": [{
                "id": _stable_id("track-audio"),
                "name": "Voiceover",
                "type": "audio",
                "elements": [audio_element],
                "muted": False,
            }],
        },
        "bookmarks": [],
        "createdAt": now,
        "updatedAt": now,
    }
    project = {
        "schema_version": 1,
        "exported_at": now,
        "project": {
            "metadata": {
                "id": project_id,
                "name": name,
                "duration": audio_dur_ticks,
                "createdAt": now,
                "updatedAt": now,
            },
            "scenes": [scene],
            "currentSceneId": scene_id,
            "settings": {
                "fps": {"numerator": 30, "denominator": 1},
                "canvasSize": CANVAS,
                "canvasSizeMode": "preset",
                "lastCustomCanvasSize": None,
                "originalCanvasSize": None,
                "background": {"type": "color", "color": "#000000"},
            },
            "version": 24,
            "timelineViewState": {"zoomLevel": 1, "scrollLeft": 0, "playheadTime": 0},
        },
        "media": manifest,
    }

    os.makedirs(os.path.dirname(os.path.abspath(out_path)), exist_ok=True)
    with open(out_path, "w") as f:
        json.dump(project, f, indent=2)

    srt_path = os.path.splitext(out_path)[0] + ".srt"
    if srt_path.endswith(".opencut.srt"):
        srt_path = srt_path[: -len(".opencut.srt")] + ".srt"
    with open(srt_path, "w") as f:
        for i, cap in enumerate(caps):
            start_s = cap["start_s"]
            end_s = caps[i + 1]["start_s"] if i + 1 < len(caps) else audio_dur_s
            f.write(f"{i + 1}\n{_srt_time(start_s)} --> {_srt_time(end_s)}\n{cap['text']}\n\n")

    return {
        "project_file": out_path,
        "srt_file": srt_path,
        "duration_s": round(audio_dur_s, 2),
        "images": len(imgs),
        "captions": len(caption_elements),
    }


def _parse_srt(text: str) -> list[dict]:
    """Parse a .srt into ordered {start, end, text} entries (ignores the index numbers)."""
    entries = []
    for block in re.split(r"\n\s*\n", text.strip()):
        lines = block.splitlines()
        if len(lines) >= 3 and "-->" in lines[1]:
            start, end = (s.strip() for s in lines[1].split("-->"))
            entries.append({"start": start, "end": end, "text": "\n".join(lines[2:]).strip()})
    return entries


def delete_captions(project_path: str, srt_path: str | None, indices: list[int]) -> dict:
    """Remove 1-based caption entries from a built OpenCut project and its .srt.

    Edits the deliverable in place: only the on-screen text overlays are dropped — the
    image and audio tracks (and the video's length) are untouched. The originals are
    backed up to <file>.bak before writing, giving a one-level undo.
    """
    with open(project_path) as f:
        project = json.load(f)

    track = None
    for scene in project.get("project", {}).get("scenes", []):
        for t in scene.get("tracks", {}).get("overlay", []):
            if t.get("type") == "text":
                track = t
                break
        if track:
            break
    if track is None or not track.get("elements"):
        raise ValueError("this project has no captions to delete")

    elements = track["elements"]
    by_time = sorted(range(len(elements)), key=lambda i: elements[i].get("startTime", 0))
    n = len(by_time)
    drop = {i for i in indices if 1 <= i <= n}
    if not drop:
        raise ValueError(f"no valid caption numbers in 1..{n}")

    keep_positions = sorted(by_time[rank - 1] for rank in range(1, n + 1) if rank not in drop)
    track["elements"] = [elements[i] for i in keep_positions]

    shutil.copyfile(project_path, project_path + ".bak")
    with open(project_path, "w") as f:
        json.dump(project, f, indent=2)

    remaining_srt = None
    if srt_path and os.path.exists(srt_path):
        with open(srt_path) as f:
            entries = _parse_srt(f.read())
        kept = [e for idx, e in enumerate(entries, start=1) if idx not in drop]
        shutil.copyfile(srt_path, srt_path + ".bak")
        with open(srt_path, "w") as f:
            for i, e in enumerate(kept, start=1):
                f.write(f"{i}\n{e['start']} --> {e['end']}\n{e['text']}\n\n")
        remaining_srt = len(kept)

    return {
        "deleted": sorted(drop),
        "remaining": len(keep_positions),
        "remaining_srt": remaining_srt,
        "project_file": project_path,
        "srt_file": srt_path,
    }

File: web/src/lib/jobs/build-project.ts
import "server-only";

import { createHash } from "node:crypto";
import { copyFileSync, existsSync, mkdirSync, readdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import path from "node:path";

import { readWav, wavDurationSeconds } from "@/lib/jobs/wav";

// Parameterized OpenCut project builder — direct port of video/api/build_project.py.
// Pure logic (JSON assembly, WAV/PNG header parsing, SRT read/write), no ML deps, so it
// ports unchanged: takes audio + transcript + images dir as arguments and assembles any
// job's artifacts into an OpenCut project file.

const TICKS_PER_SECOND = 120_000;
const CANVAS = { width: 1920, height: 1080 };
const NS = "6f0c1e00-0000-4000-8000-000000000001"; // fixed namespace → stable ids

const IMG_RE = /^(\d+)_(\d\d)-(\d\d)\.png$/;
const CAP_RE = /^\[(\d{1,2}):(\d{2})\]\s*(.+?)\s*$/;

function uuidToBytes(uuid: string): Buffer {
  return Buffer.from(uuid.replace(/-/g, ""), "hex");
}

function bytesToUuid(bytes: Buffer): string {
  const hex = bytes.toString("hex");
  return `${hex.slice(0, 8)}-${hex.slice(8, 12)}-${hex.slice(12, 16)}-${hex.slice(16, 20)}-${hex.slice(20, 32)}`;
}

// RFC 4122 v5 (name-based, SHA-1) — matches Python's uuid.uuid5, so re-building the
// same job's project twice yields the same media/element ids.
function uuidv5(namespace: string, name: string): string {
  const hash = createHash("sha1")
    .update(Buffer.concat([uuidToBytes(namespace), Buffer.from(name, "utf8")]))
    .digest();
  const bytes = Buffer.from(hash.subarray(0, 16));
  bytes[6] = (bytes[6] & 0x0f) | 0x50;
  bytes[8] = (bytes[8] & 0x3f) | 0x80;
  return bytesToUuid(bytes);
}

const stableId = (name: string) => uuidv5(NS, name);
const ticks = (seconds: number) => Math.round(seconds * TICKS_PER_SECOND);

function pngSize(filePath: string): { width: number; height: number } {
  const fd = readFileSync(filePath);
  const head = fd.subarray(0, 24);
  return { width: head.readUInt32BE(16), height: head.readUInt32BE(20) };
}

function wavDuration(filePath: string): number {
  return wavDurationSeconds(readWav(readFileSync(filePath)));
}

function srtTime(seconds: number): string {
  const ms = Math.round(seconds * 1000);
  const h = Math.floor(ms / 3_600_000);
  const m = Math.floor((ms % 3_600_000) / 60_000);
  const s = Math.floor((ms % 60_000) / 1000);
  const rem = ms % 1000;
  const pad = (n: number, len = 2) => String(n).padStart(len, "0");
  return `${pad(h)}:${pad(m)}:${pad(s)},${pad(rem, 3)}`;
}

export type BuildProjectArgs = {
  audioFile: string;
  imagesDir: string;
  outPath: string;
  transcriptFile?: string | null;
  name?: string;
  includeCaptions?: boolean;
};

export type BuildSummary = {
  project_file: string;
  srt_file: string;
  duration_s: number;
  images: number;
  captions: number;
};

export function buildProject({
  audioFile,
  imagesDir,
  outPath,
  transcriptFile,
  name = "Generated Video",
  includeCaptions = false,
}: BuildProjectArgs): BuildSummary {
  const audioDurS = wavDuration(audioFile);
  const audioDurTicks = ticks(audioDurS);
  const audioName = path.basename(audioFile);
  const audioMediaId = stableId(audioName);

  const imgs = readdirSync(imagesDir)
    .sort()
    .map((n) => {
      const m = IMG_RE.exec(n);
      if (!m) return null;
      return { name: n, startS: Number(m[2]) * 60 + Number(m[3]) };
    })
    .filter((x): x is { name: string; startS: number } => x !== null);
  if (imgs.length === 0) {
    throw new Error(`no timestamped PNGs (NNN_MM-SS.png) found in ${imagesDir}`);
  }

  const imageElements = [];
  const manifest: Record<string, unknown>[] = [
    {
      mediaId: audioMediaId,
      filename: audioName,
      type: "audio",
      size: statSync(audioFile).size,
      duration: Math.round(audioDurS * 1000) / 1000,
    },
  ];
  for (let i = 0; i < imgs.length; i++) {
    const img = imgs[i];
    const startS = img.startS;
    const endS = i + 1 < imgs.length ? imgs[i + 1].startS : audioDurS;
    const startT = ticks(startS);
    const durT = ticks(endS) - startT;
    const mid = stableId(img.name);
    const imgPath = path.join(imagesDir, img.name);
    const { width, height } = pngSize(imgPath);
    imageElements.push({
      id: stableId(`el-${img.name}`),
      name: img.name,
      type: "image",
      mediaId: mid,
      startTime: startT,
      duration: durT,
      trimStart: 0,
      trimEnd: 0,
      transform: { scaleX: 1, scaleY: 1, position: { x: 0, y: 0 }, rotate: 0 },
      opacity: 1,
    });
    manifest.push({
      mediaId: mid,
      filename: img.name,
      type: "image",
      size: statSync(imgPath).size,
      width,
      height,
    });
  }

  const caps: { startS: number; text: string }[] = [];
  if (transcriptFile && existsSync(transcriptFile)) {
    for (const line of readFileSync(transcriptFile, "utf8").split("\n")) {
      const m = CAP_RE.exec(line);
      if (m) caps.push({ startS: Number(m[1]) * 60 + Number(m[2]), text: m[3] });
    }
  }

  const captionElements = [];
  if (includeCaptions) {
    for (let i = 0; i < caps.length; i++) {
      const cap = caps[i];
      const startS = cap.startS;
      const endS = i + 1 < caps.length ? caps[i + 1].startS : audioDurS;
      const startT = ticks(startS);
      const durT = Math.max(ticks(endS) - startT, 1);
      captionElements.push({
        id: stableId(`cap-${i}`),
        name: cap.text.slice(0, 40),
        type: "text",
        content: cap.text,
        fontSize: 6,
        fontFamily: "Arial",
        color: "#ffffff",
        background: {
          enabled: true,
          color: "rgba(0,0,0,0.6)",
          cornerRadius: 8,
          paddingX: 24,
          paddingY: 12,
          offsetX: 0,
          offsetY: 0,
        },
        textAlign: "center",
        fontWeight: "bold",
        fontStyle: "normal",
        textDecoration: "none",
        letterSpacing: 0,
        lineHeight: 1.2,
        startTime: startT,
        duration: durT,
        trimStart: 0,
        trimEnd: 0,
        transform: { scaleX: 1, scaleY: 1, position: { x: 0, y: 330 }, rotate: 0 },
        opacity: 1,
      });
    }
  }

  const audioElement = {
    id: stableId("el-audio"),
    name: audioName,
    type: "audio",
    sourceType: "upload",
    mediaId: audioMediaId,
    volume: 1,
    startTime: 0,
    duration: audioDurTicks,
    trimStart: 0,
    trimEnd: 0,
    sourceDuration: audioDurTicks,
  };

  const now = new Date().toISOString();
  const sceneId = stableId("scene-main");
  const projectId = stableId(`project-${name}`);
  const scene = {
    id: sceneId,
    name: "Main scene",
    isMain: true,
    tracks: {
      overlay:
        includeCaptions && captionElements.length > 0
          ? [{ id: stableId("track-captions"), name: "Captions", type: "text", elements: captionElements, hidden: false }]
          : [],
      main: {
        id: stableId("track-main"),
        name: "Main Track",
        type: "video",
        elements: imageElements,
        muted: false,
        hidden: false,
      },
      audio: [{ id: stableId("track-audio"), name: "Voiceover", type: "audio", elements: [audioElement], muted: false }],
    },
    bookmarks: [],
    createdAt: now,
    updatedAt: now,
  };
  const project = {
    schema_version: 1,
    exported_at: now,
    project: {
      metadata: { id: projectId, name, duration: audioDurTicks, createdAt: now, updatedAt: now },
      scenes: [scene],
      currentSceneId: sceneId,
      settings: {
        fps: { numerator: 30, denominator: 1 },
        canvasSize: CANVAS,
        canvasSizeMode: "preset",
        lastCustomCanvasSize: null,
        originalCanvasSize: null,
        background: { type: "color", color: "#000000" },
      },
      version: 24,
      timelineViewState: { zoomLevel: 1, scrollLeft: 0, playheadTime: 0 },
    },
    media: manifest,
  };

  mkdirSync(path.dirname(path.resolve(outPath)), { recursive: true });
  writeFileSync(outPath, JSON.stringify(project, null, 2));

  const ext = path.extname(outPath);
  let srtPath = outPath.slice(0, outPath.length - ext.length) + ".srt";
  if (srtPath.endsWith(".opencut.srt")) srtPath = srtPath.slice(0, -".opencut.srt".length) + ".srt";
  const srtLines: string[] = [];
  for (let i = 0; i < caps.length; i++) {
    const cap = caps[i];
    const startS = cap.startS;
    const endS = i + 1 < caps.length ? caps[i + 1].startS : audioDurS;
    srtLines.push(`${i + 1}\n${srtTime(startS)} --> ${srtTime(endS)}\n${cap.text}\n`);
  }
  writeFileSync(srtPath, srtLines.join("\n"));

  return {
    project_file: outPath,
    srt_file: srtPath,
    duration_s: Math.round(audioDurS * 100) / 100,
    images: imgs.length,
    captions: captionElements.length,
  };
}

function parseSrt(text: string): { start: string; end: string; text: string }[] {
  const entries: { start: string; end: string; text: string }[] = [];
  for (const block of text.trim().split(/\n\s*\n/)) {
    const lines = block.split("\n");
    if (lines.length >= 3 && lines[1].includes("-->")) {
      const [start, end] = lines[1].split("-->").map((s) => s.trim());
      entries.push({ start, end, text: lines.slice(2).join("\n").trim() });
    }
  }
  return entries;
}

export type DeleteCaptionsSummary = {
  deleted: number[];
  remaining: number;
  remaining_srt: number | null;
  project_file: string;
  srt_file: string | null;
};

/**
 * Remove 1-based caption entries from a built OpenCut project and its .srt. Edits the
 * deliverable in place: only the text overlay track is touched — images/audio and the
 * video's length are untouched. Originals are backed up to <file>.bak (one-level undo).
 */
export function deleteCaptions(
  projectPath: string,
  srtPath: string | null | undefined,
  indices: number[],
): DeleteCaptionsSummary {
  const project = JSON.parse(readFileSync(projectPath, "utf8"));

  type OverlayTrack = { type: string; elements: { startTime?: number }[] };
  let track: OverlayTrack | null = null;
  for (const scene of project?.project?.scenes ?? []) {
    for (const t of scene?.tracks?.overlay ?? []) {
      if (t.type === "text") {
        track = t;
        break;
      }
    }
    if (track) break;
  }
  if (!track || !track.elements || track.elements.length === 0) {
    throw new Error("this project has no captions to delete");
  }

  const elements = track.elements;
  const byTime = elements.map((_, i) => i).sort((a, b) => (elements[a].startTime ?? 0) - (elements[b].startTime ?? 0));
  const n = byTime.length;
  const drop = new Set(indices.filter((i) => i >= 1 && i <= n));
  if (drop.size === 0) throw new Error(`no valid caption numbers in 1..${n}`);

  const keepPositions = Array.from({ length: n }, (_, i) => i + 1)
    .filter((rank) => !drop.has(rank))
    .map((rank) => byTime[rank - 1])
    .sort((a, b) => a - b);
  track.elements = keepPositions.map((i) => elements[i]);

  copyFileSync(projectPath, `${projectPath}.bak`);
  writeFileSync(projectPath, JSON.stringify(project, null, 2));

  let remainingSrt: number | null = null;
  if (srtPath && existsSync(srtPath)) {
    const entries = parseSrt(readFileSync(srtPath, "utf8"));
    const kept = entries.filter((_, idx) => !drop.has(idx + 1));
    copyFileSync(srtPath, `${srtPath}.bak`);
    const lines = kept.map((e, i) => `${i + 1}\n${e.start} --> ${e.end}\n${e.text}\n`);
    writeFileSync(srtPath, lines.join("\n"));
    remainingSrt = kept.length;
  }

  return {
    deleted: Array.from(drop).sort((a, b) => a - b),
    remaining: keepPositions.length,
    remaining_srt: remainingSrt,
    project_file: projectPath,
    srt_file: srtPath ?? null,
  };
}
